lib/luaxtable.c

```diff
@@ -32,48 +14,170 @@
 {
-	return 0;
+	if (lunatik_getregistry(L, xtable->skb) != LUA_TUSERDATA)
+		return NULL;
+
+	return (lunatik_object_t *)lunatik_toobject(L, -1);
 }
 
-static void luaxtable_match_destroy(const struct xt_mtdtor_param *par)
+static inline void luaxtable_pushparams(lua_State *L, const struct xt_action_param *par)
 {
+	lua_newtable(L);
+	lua_pushboolean(L, par->hotdrop);
+	lua_setfield(L, -2, "hotdrop");
+	lua_pushinteger(L, par->thoff);
+	lua_setfield(L, -2, "thoff");
+	lua_pushinteger(L, par->fragoff);
+	lua_setfield(L, -2, "fragoff");
+#if (LINUX_VERSION_CODE >= KERNEL_VERSION(4, 10, 0))
+	lua_pushinteger(L, xt_hooknum(par));
+#else
+	lua_pushinteger(L, par->hooknum);
+#endif
+	lua_setfield(L, -2, "hooknum");
+	lua_pushvalue(L, -1); /* param table */
+	lua_insert(L, lua_gettop(L) - 2); /* stack: param, skb, param */
 }
 
-static unsigned int luaxtable_target(struct sk_buff *skb, const struct xt_action_param *par)
+static inline int luaxtable_pushskb(lua_State *L, luaxtable_t *xtable, const struct sk_buff *skb, uint8_t opt)
 {
-	return NF_ACCEPT;
+	lunatik_object_t *data = luaxtable_getskb(L, xtable);
+	if (data == NULL) {
+		pr_err("could not find skb\n");
+		return -1;
+	}
+
+	luadata_reset(data, skb->data, skb_headlen(skb), opt);
+	return 0;
 }
 
-static int luaxtable_target_check(const struct xt_tgchk_param *par)
+static int luaxtable_domatch(lua_State *L, luaxtable_t *xtable, const struct sk_buff *skb, struct xt_action_param *par)
 {
-	return 0;
+	int ret;
+	if (luaxtable_pushskb(L, xtable, skb, LUADATA_OPT_READONLY) == -1)
+		return luaxtable_hooks.match_fallback;
+	luaxtable_pushparams(L, par);
+
+	ret = luaxtable_invoke(L, xtable, "match", 2, 1, luaxtable_hooks.match_fallback);
+	lua_getfield(L, -2, "hotdrop");
+	par->hotdrop = lua_toboolean(L, -1);
+	return ret;
 }
 
-static void luaxtable_target_destroy(const struct xt_tgdtor_param *par)
+static int luaxtable_dotarget(lua_State *L, luaxtable_t *xtable, struct sk_buff *skb, const struct xt_action_param *par)
 {
+	if (luaxtable_pushskb(L, xtable, skb, LUADATA_OPT_NONE) == -1)
+		return luaxtable_hooks.match_fallback;
+	luaxtable_pushparams(L, par);
+	return luaxtable_invoke(L, xtable, "target", 2, 1, luaxtable_hooks.target_fallback);
 }
 
+#define LUAXTABLE_HOOK_CB(T, U, V, hook, huk) 				\
+static T luaxtable_##hook(U skb, V par)					\
+{									\
+	int ret;							\
+	const luaxtable_info_t *info = (const luaxtable_info_t *)par->huk##info;	\
+	luaxtable_t *xtable = info->data;				\
+									\
+	lunatik_run(xtable->runtime, luaxtable_do##hook, ret, xtable, skb, par);	\
+	return ret;							\
+}
+
+#define LUAXTABLE_CHECKER_CB(hook, hk, huk, HOOK)				\
+static int luaxtable_##hook##_check(const struct xt_##hk##chk_param *par)	\
+{										\
+	int ret;								\
+	luaxtable_t *xtable;							\
+	lunatik_object_t *obj = luarcu_gettable(luaxtable_hooks.hook, par->hook->name, XT_EXTENSION_MAXNAMELEN);	\
+	if (obj == NULL) {							\
+		pr_err("could not find table for %s\n", par->hook->name);	\
```

Comment:
which table do you refer here? `xtable`? `rcu`?

Reply:
`rcu` table

Reply:
the API client has no clue of the RCU table, right? it's transparent to them.. I would change this error message..

Reply:
What can be a better message here ?

Reply:
perhaps..
```suggestion
		pr_err("could not find hook (%s)\n", par->hook->name);	\
```

didn't like that much, but I could think on something much better.. =)